### ISSUE TITLE
Custom scatter function for DataParallel

### ISSUE THREAD

**Issue by CorentinJ** [pytorch/pytorch#52597]
## 🚀 Feature
Allow for passing an optional argument `scatter_fn` to `nn.DataParallel` that defaults to `nn.parallel.scatter_gather.scatter`. This function would have a signature match `nn.parallel.scatter_gather.scatter` and would be called instead when provided.

## Motivation
I've encountered issues with the base scatter function when passing custom tensor data objects with `DataParallel`. These include:
- Passing a namedtuple of tensors: items were dispatched individually to replicas when they were intended to replicated on each.
- Passing a custom collection of tensors that is a subclass of `dict`: the constructor was not matching that of `dict`, causing bugs.

## Pitch
This change would allow for more flexibility towards custom tensor collections in use with `DataParallel`. In my experience, such custom objects become strongly desirable when working on a large-scale project.

**Comment by SsnL** [comment]
one workaround is to subclass `DataParallel` and override `scatter`

**Comment by CorentinJ** [comment]
> one workaround is to subclass `DataParallel` and override `scatter`

Yes, although `nn.DataParallel.scatter` calls `nn.parallel.scatter_gather.scatter_kwargs` which itself calls `nn.parallel.scatter_gather.scatter`. I'm unsure what the purpose of `scatter_kwargs` is, but if it can be safely replaced by a direct call to the `scatter` function then that would do the trick. Is that the case?
